book-2-platform-api/chapters/API_MODELS_CONTROLLERS.md

```diff
@@ -96,70 +158,121 @@ namespace CoffeeShop.Controllers
+                        {
+                            Id = reader.GetInt32(reader.GetOrdinal("Id")),
+                            Title = reader.GetString(reader.GetOrdinal("Title")),
+                            BeanType = reader.GetString(reader.GetOrdinal("BeanType"))
+                        };
+
+                        coffees.Add(coffee);
+                    }
+                    reader.Close();
 
+                    return Ok(coffees);
+                }
+            }
         }
 
         [HttpGet("{id}", Name = "GetCoffee")]
         public async Task<IActionResult> Get([FromRoute] int id)
         {
-            using (IDbConnection conn = Connection)
+            using (SqlConnection conn = Connection)
             {
-                string sql = $"SELECT * FROM Coffee WHERE Id = {id}";
+                conn.Open();
+                using (SqlCommand cmd = conn.CreateCommand())
+                {
+                    cmd.CommandText = @"
+                        SELECT
+                            Id, Title, BeanType
+                        FROM Coffee
+                        WHERE Id = @id";
+                    cmd.Parameters.Add(new SqlParameter("@id", id));
+                    SqlDataReader reader = cmd.ExecuteReader();
 
-                var singleCoffee = (await conn.QueryAsync<Coffee>(sql)).Single();
-                return Ok(singleCoffee);
+                    Coffee coffee = null;
+
+                    if (reader.Read())
+                    {
+                        coffee = new coffee
+                        {
+                            Id = reader.GetInt32(reader.GetOrdinal("Id")),
+                            Title = reader.GetString(reader.GetOrdinal("Title")),
+                            BeanType = reader.GetString(reader.GetOrdinal("BeanType"))
+                        };
+                    }
+                    reader.Close();
+
+                    return Ok(coffee);
+                }
             }
         }
 
         [HttpPost]
         public async Task<IActionResult> Post([FromBody] Coffee coffee)
         {
-            string sql = $@"INSERT INTO Coffee
-            (Title, BeanType)
-            VALUES
-            ('{coffee.Title}', '{coffee.BeanType}');
-            select MAX(Id) from Coffee";
-
-            using (IDbConnection conn = Connection)
+            using (SqlConnection conn = Connection)
             {
-                var newId = (await conn.QueryAsync<int>(sql)).Single();
-                coffee.Id = newId;
-                return CreatedAtRoute("GetCoffee", new { id = newId }, coffee);
+                conn.Open();
+                using (SqlCommand cmd = conn.CreateCommand())
+                {
+                    cmd.CommandText = @"INSERT INTO Coffee (Title, BeanType)
+                                        OUTPUT INSERTED.Id
+                                        VALUES (@title, @beanType);
+                                        select MAX(Id) from Coffee";
```

Comment:
selecting the max id is usually but not always right.  consider using `@@identity` or `scope_identity()` for scoping to the current connection and query.

Reply:
The question becomes "what will the students understand?" or "Do they even need to understand this?"

I've used `scope_identity()`  past versions of our exercise and group project code, but @askingalot brought up a valid concern that students may get confused by seeing both an INSERT and a SELECT in the same command.  Having the `OUTPUT` clause there in the SQL might help them understand the the insert returns something.

Open to discussion.

Reply:
Ah Ok, I missed the `Output` clause because I was looking at the select, I think what you need to do then is remove the select entirely, its not necessary.

Reply:
Doh! Missed that. Thanks